src/optimizer/plugins/SimplifyArrayIndexingOptimizer.ts

```diff
@@ -0,0 +1,123 @@
+          // Add a prefix to the current element containing previously parsed rule elements
+          element.prefix = utils.assembleFSHPath(parsedRule.path.slice(0, elementIndex));
+          applySoftIndexing(element, pathMap);
+        });
+
+        parsedRule.caretPath?.forEach((element: fhirtypes.PathPart, elementIndex: number) => {
+          // Caret path indexes should only be resolved in the context of a specific path
+          // Each normal path has a separate map to keep track of the caret path indexes
+          if (!caretPathMap.has(originalRule.path)) {
+            caretPathMap.set(originalRule.path, new Map());
+          }
+
+          const elementCaretPathMap = caretPathMap.get(originalRule.path);
+          // Add a prefix to the current element containing previously parsed rule elements
+          element.prefix = utils.assembleFSHPath(parsedRule.caretPath.slice(0, elementIndex));
+          applySoftIndexing(element, elementCaretPathMap);
+        });
+      });
+
+      removeZeroIndices();
+
+      parsedRules.forEach((parsedRule: any, ruleIndex: number) => {
+        const originalRule = def.rules[ruleIndex];
+        originalRule.path = utils.assembleFSHPath(parsedRule.path);
+        if (originalRule instanceof ExportableCaretValueRule) {
+          originalRule.caretPath = utils.assembleFSHPath(parsedRule.caretPath);
+        }
+      });
+    });
+  }
+} as OptimizerPlugin;
+
+/**
+ *
+ * @param {PathPart} element - A single element in a rules path
+ * @param {Map<string, number} pathMap - A map containing an element's name as the key and that element's updated index as the value
+ */
+function applySoftIndexing(element: fhirtypes.PathPart, pathMap: Map<string, number>) {
+  // Must account for a pathPart's base name, prior portions of the path, as well as any slices it's contained in.
+  const mapName = `${element.prefix ?? ''}.${element.base}|${(element.slices ?? []).join('|')}`;
+  const indexRegex = /^[0-9]+$/;
+
+  const currentNumericBracket = element.brackets?.find(bracket => indexRegex.test(bracket));
+  if (!currentNumericBracket) return; // If the element is not an array, exit the function
+
+  const bracketIndex = element.brackets.indexOf(currentNumericBracket);
+
+  if (!pathMap.has(mapName)) {
+    pathMap.set(mapName, parseInt(currentNumericBracket));
+    // If the element contains a zero Index, we'll flag the rule to have the index removed
+    if (currentNumericBracket === '0') singletonArrayElements.set(mapName, [element]);
+  } else {
+    const previousNumericIndex = pathMap.get(mapName);
+    if (parseInt(currentNumericBracket) === previousNumericIndex + 1) {
+      if (previousNumericIndex === 0) singletonArrayElements.delete(mapName);
+      element.brackets[bracketIndex] = '+';
+      pathMap.set(mapName, previousNumericIndex + 1);
+    } else if (parseInt(currentNumericBracket) === previousNumericIndex) {
```

Comment:
I think I am running into some tricky issues with how this `pathMap` setup deals with nested paths. I've been trying it out on the US Core Client Capability Statement resource. The first time we see this path:
```
* rest[0].resource[0].supportedProfile[0] = "http://hl7.org/fhir/us/core/StructureDefinition/us-core-allergyintolerance"
```
It is correctly translated into:
```
* rest.resource[=].supportedProfile[0] = "http://hl7.org/fhir/us/core/StructureDefinition/us-core-allergyintolerance"
```
But then the second time we run into a path ending in `supportedProfile`:
```
* rest[0].resource[1].supportedProfile[0] = "http://hl7.org/fhir/us/core/StructureDefinition/us-core-careplan"
```
It gets translated into
```
* rest.resource[=].supportedProfile[=] = "http://hl7.org/fhir/us/core/StructureDefinition/us-core-careplan"
```
Which isn't really right, and causes SUSHI to emit an error like:
![image](https://user-images.githubusercontent.com/52747882/111827530-61275c80-88c0-11eb-9d50-dba90c6f4794.png)

It is correct to assume the `0` index, but still we don't want to write FSH that produces a bunch of errors. From what I can tell this happens because the second time we hit a path ending in `supportedProfile`, we already have `rest[=].resource[=].supportedProfile|` in the `pathMap`. But _this_ path really has nothing to do with that original `supportedProfile`. The `pathMap` does not account for the fact that the index of `resource` has been incremented since the first time we saw `supportedProfile`. At least I think that is what is happening. If I am way off base on my explanation let me know. But this is definitely a bug, regardless of if I've explained it all wrong, since it is causing a lot of errors when I try to round trip US Core.

Reply:
Ahhhh I see what's going on here. Cases like this are supposed to be caught by adding the `prefix` property to each PathPart in a path. Problem was, this property was being added _after_ each property's numeric indexes were converted to soft indexes. This can create situations like what you ran into here, thanks for catching this! This should be addressed in the next commit.

Reply:
Addressed in [`d6e165e`](https://github.com/FHIR/GoFSH/pull/106/commits/d6e165e5611b739e6fae61fe8ba1ca39183768ae).